daemon/cmd/endpoint.go

```diff
@@ -190,6 +192,93 @@ func (d *Daemon) errorDuringCreation(ep *endpoint.Endpoint, err error) (*endpoin
 	return nil, PutEndpointIDFailedCode, err
 }
 
+type endpointCreationRequest struct {
+	// cancel is the cancellation function that can be called to cancel
+	// this endpoint create request
+	cancel context.CancelFunc
+
+	// endpoint is the endpoint being added in the request
+	endpoint *endpoint.Endpoint
+
+	// started is the timestamp on when the processing has started
+	started time.Time
+}
+
+type endpointCreationManager struct {
+	mutex    lock.Mutex
+	requests map[string]*endpointCreationRequest
+}
+
+func newEndpointCreationManager() *endpointCreationManager {
+	return &endpointCreationManager{
+		requests: map[string]*endpointCreationRequest{},
+	}
+}
+
+func (m *endpointCreationManager) NewCreateRequest(ep *endpoint.Endpoint, cancel context.CancelFunc) {
+	// Tracking is only performed if Kubernetes pod names are available.
+	// The endpoint create logic already ensures that IPs and containerID
+	// are unique and thus tracking is not required outside of the
+	// Kubernetes context
+	if !ep.K8sNamespaceAndPodNameIsSet() || !k8s.IsEnabled() {
+		return
+	}
+
+	podName := ep.GetK8sNamespaceAndPodName()
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if req, ok := m.requests[podName]; ok {
+		ep.Logger(daemonSubsys).Warning("Cancelling obsolete endpoint creating due to new create for same pod")
+		req.cancel()
+	}
+
+	ep.Logger(daemonSubsys).Debug("New create request")
+	m.requests[podName] = &endpointCreationRequest{
+		cancel:   cancel,
+		endpoint: ep,
+		started:  time.Now(),
+	}
+}
+
+func (m *endpointCreationManager) EndCreateRequest(ep *endpoint.Endpoint) bool {
+	if !ep.K8sNamespaceAndPodNameIsSet() || !k8s.IsEnabled() {
+		return false
+	}
+
+	podName := ep.GetK8sNamespaceAndPodName()
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if req, ok := m.requests[podName]; ok {
+		if req.endpoint == ep {
```

Comment:
In the "else" case of this condition, what do we expect? Is that a bug? Should we warn? That'd basically be a duplicate pod namespace+name currently being created which the rest of this PR seems to assume is not possible AFAICT.

Reply:
Names have to be unique, can disregard: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/